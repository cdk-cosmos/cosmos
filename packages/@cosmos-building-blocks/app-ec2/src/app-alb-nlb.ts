/* eslint-disable @typescript-eslint/camelcase */
import { CfnOutput, Construct, Fn } from '@aws-cdk/core';
import {
  CfnVPCEndpoint,
  CfnVPCEndpointService,
  IInterfaceVpcEndpoint,
  InterfaceVpcEndpoint,
  IVpc,
  IVpcEndpointServiceLoadBalancer,
  SecurityGroup,
  SubnetType,
  VpcEndpointService,
} from '@aws-cdk/aws-ec2';
import {
  NetworkLoadBalancer,
  NetworkListener,
  NetworkTargetGroup,
  ApplicationLoadBalancer,
  CfnListener,
  TargetType,
} from '@aws-cdk/aws-elasticloadbalancingv2';
import { Rule, Schedule, IRuleTarget, RuleTargetInput, IRule } from '@aws-cdk/aws-events';
import { ArnPrincipal } from '@aws-cdk/aws-iam';
import { IFunction } from '@aws-cdk/aws-lambda';
import { CfnRecordSet } from '@aws-cdk/aws-route53';
import { Bucket } from '@aws-cdk/aws-s3';
import { StringParameter } from '@aws-cdk/aws-ssm';

export interface AppProps {
  appAlbNlbVpc: IVpc;
  envName: string;
}

export interface AppAlbProps extends AppProps {
  appAlbNlbVpc: IVpc;
  albSecGroupId: string;
  bucketPrefix: string;
  certificateArn: string;
  envName: string;
  logBucket: string;
}

export interface AppNlbComboCloudwatchLambdaRuleProps {
  lambdaForUpdatingTargets: IFunction;
  scheduleRate?: string;
  // targetInputText: string;
  // eslint unhappy about any but that is the type the cdk expects
  // eslint-disable-next-line
  targetInputObject: any;
}

export interface AppNlbComboInterfaceVpcEndpointProps {
  envName: string;
  exportName?: string;
  port?: number;
  subnetType?: SubnetType;
  vpc: IVpc;
  vpcEndpointServiceLoadBalancers: IVpcEndpointServiceLoadBalancer[];
  whitelistedPrincipals?: ArnPrincipal[];
}

export interface NlbEndpointComboProps extends AppProps {
  albDnsAddress: string;
  lambdaForUpdatingTargets: IFunction;
  nlbEndpointVpc?: IVpc;
  parameterPath?: string;
  publicHostedZoneId?: string;
  recordName?: string;
  recordSetPrefix?: string;
  zoneName?: string;
}

export class AppNlbComboCloudwatchLambdaRule extends Construct {
  public readonly rule: IRule;

  constructor(scope: Construct, id: string, props: AppNlbComboCloudwatchLambdaRuleProps) {
    super(scope, id);

    const t1: IRuleTarget = {
      bind: () => ({
        id: '',
        arn: props.lambdaForUpdatingTargets.functionArn,
        input: RuleTargetInput.fromObject(props.targetInputObject),
      }),
    };

    const scheduleRate = props.scheduleRate || 'rate(180 minutes)';

    this.rule = new Rule(this, id, {
      // Name
      // Description
      schedule: Schedule.expression(scheduleRate),
      targets: [t1],
    });
  }
}

export class AppAlb extends Construct {
  public readonly albDnsAddress: string;
  public readonly listener: CfnListener;
  public readonly internalListener: CfnListener;

  constructor(scope: Construct, id: string, props: AppAlbProps) {
    super(scope, id);

    const vpc = props.appAlbNlbVpc;

    const albSecurityGroup = SecurityGroup.fromSecurityGroupId(this, 'sg', props.albSecGroupId);

    // Application Load Balancers
    const uiAlb = new ApplicationLoadBalancer(this, `${props.envName}-alb`, {
      vpc,
      internetFacing: false,
      loadBalancerName: `${props.envName}-alb`,
      deletionProtection: false,
      // vpcSubnets: subnets,
      securityGroup: albSecurityGroup,
    });

    const logBucket = Bucket.fromBucketName(this, 'logBucket', props.logBucket);
    uiAlb.logAccessLogs(logBucket, props.bucketPrefix);

    // const action = new CfnListener.ActionProperty();
    const listener = new CfnListener(this, 'listener', {
      defaultActions: [
        {
          fixedResponseConfig: {
            contentType: 'text/plain',
            messageBody: 'hi from the default last rule',
            statusCode: '200',
          },
          type: 'fixed-response',
        },
      ],
      loadBalancerArn: uiAlb.loadBalancerArn,
      port: 443,
      protocol: 'HTTPS',
      certificates: [{ certificateArn: props.certificateArn }],
    });

    // const action = new CfnListener.ActionProperty();
    const internalListener = new CfnListener(this, 'InternalListener', {
      defaultActions: [
        {
          fixedResponseConfig: {
            contentType: 'text/plain',
            messageBody: 'hi from the default last rule',
            statusCode: '200',
          },
          type: 'fixed-response',
        },
      ],
      loadBalancerArn: uiAlb.loadBalancerArn,
      port: 9443,
      protocol: 'HTTPS',
      certificates: [{ certificateArn: props.certificateArn }],
    });

    new CfnOutput(this, 'ListenerOutput', {
      value: listener.ref,
      exportName: `${props.envName}-Listener`,
    });

    this.listener = listener;

    new CfnOutput(this, 'InternalListenerOutput', {
      value: internalListener.ref,
      exportName: `${props.envName}-InternalListener`,
    });

    this.internalListener = internalListener;

    this.albDnsAddress = uiAlb.loadBalancerDnsName;
  }
}

export class AppNlbComboInterfaceVpcEndpoint extends Construct {
  public readonly interfaceVpcEndpoint: IInterfaceVpcEndpoint;
  public readonly vpcEndpointId: string;
  public readonly endpointDns: string;

  constructor(scope: Construct, id: string, props: AppNlbComboInterfaceVpcEndpointProps) {
    super(scope, id);

    // default is to allow root account to use endpoint
    const whitelistedPrincipals = props.whitelistedPrincipals || [
      new ArnPrincipal(Fn.sub('arn:aws:iam::${AWS::AccountId}:root')),
    ];

    const vpcEndpointService = new VpcEndpointService(this, 'vpcEndpointService', {
      acceptanceRequired: false,
      vpcEndpointServiceLoadBalancers: props.vpcEndpointServiceLoadBalancers,
      whitelistedPrincipals,
    });

    const cfnVPCEndpointService = vpcEndpointService.node.findChild('vpcEndpointService') as CfnVPCEndpointService;

    const subnetType = props.subnetType || SubnetType.PRIVATE;
    const port = props.port || 443;

    this.interfaceVpcEndpoint = new InterfaceVpcEndpoint(this, 'endpoint', {
      vpc: props.vpc,
      service: {
        name: Fn.sub('com.amazonaws.vpce.${AWS::Region}.${serviceName}', {
          serviceName: cfnVPCEndpointService.ref,
        }),
        port,
      },
      privateDnsEnabled: false,
      subnets: {
        subnetType,
      },
    });

    const exportName = props.exportName || 'NlbVpcEndpointId';

    new CfnOutput(this, 'NlbVpcEndpointId', {
      value: this.interfaceVpcEndpoint.vpcEndpointId,
      exportName: `${props.envName}-${exportName}`,
    });

    this.vpcEndpointId = this.interfaceVpcEndpoint.vpcEndpointId;

    const cfnVpcEndpoint = this.interfaceVpcEndpoint.node.findChild('Resource') as CfnVPCEndpoint;

    new CfnOutput(this, 'ListenerOutput', {
      value: Fn.select(1, Fn.split(':', Fn.select(0, cfnVpcEndpoint.attrDnsEntries))),
      exportName: `${props.envName}-EndpointDns`,
    });

    this.endpointDns = Fn.select(1, Fn.split(':', Fn.select(0, cfnVpcEndpoint.attrDnsEntries)));
  }
}

export class NlbEndpointCombo extends Construct {
  constructor(scope: Construct, id: string, props: NlbEndpointComboProps) {
    super(scope, id);

    const vpc = props.appAlbNlbVpc;

    // Default is to use same vpc
    let remoteEndpointVpc = props.appAlbNlbVpc;

    if (props.nlbEndpointVpc !== undefined) {
      remoteEndpointVpc = props.nlbEndpointVpc;
    }

    // Network Load Balancers
    // TODO: subnets, sg
    const netAlb = new NetworkLoadBalancer(this, `${props.envName}-nlb`, {
      vpc,
      // TODO: props
      internetFacing: false,
      loadBalancerName: `${props.envName}-nlb`,
      deletionProtection: false,
      // TODO: props
      crossZoneEnabled: true,
      // vpcSubnets: subnets,
      // securityGroup: albSecurityGroup,
    });

    // AccessLogs only for TLS
    // const logBucket = s3.Bucket.fromBucketName(this, 'logBucket', props.logBucket);
    // netAlb.logAccessLogs(logBucket, props.bucketPrefix);

    const networkTargetGroup = new NetworkTargetGroup(this, `${props.envName}-nlb-tg`, {
      port: 443,
      vpc,
      targetType: TargetType.IP,
    });

    new NetworkListener(this, `${props.envName}-nlb-listener`, {
      loadBalancer: netAlb,
      port: 443,
      defaultTargetGroups: [networkTargetGroup],
    });

    const parameterPath = props.parameterPath || '/App/Alb/app-nlb/';
    const parameterName = `${parameterPath}${props.envName}/dns-saved`;

    // Create a new SSM Parameter holding a String
    new StringParameter(this, 'StringParameter', {
      parameterName,
      stringValue: 'x',
    });

    const inputText = {
      alb_dns_address: props.albDnsAddress,
      alb_dns_parameter: parameterName,
      debug: true,
      dry_run: false,
      alb_target_group_arn: networkTargetGroup.targetGroupArn,
    };

    new AppNlbComboCloudwatchLambdaRule(this, 'appAlbNlbRule', {
      lambdaForUpdatingTargets: props.lambdaForUpdatingTargets,
      // targetInputText: JSON.stringify(inputText),
      targetInputObject: inputText,
    });

    const appNlbComboInterfaceVpcEndpoint = new AppNlbComboInterfaceVpcEndpoint(this, 'appNlbEndpoint', {
      vpc: remoteEndpointVpc,
      envName: props.envName,
      vpcEndpointServiceLoadBalancers: [netAlb],
    });

    // If you define a zone we can publish the dns - otherwise leave it up to the app to do this bit
    if ((props.publicHostedZoneId && !props.zoneName) || (!props.publicHostedZoneId && props.zoneName)) {
      throw new Error('both publicHostedZoneId and zoneName are needed or else both should be empty');
    } else if (props.publicHostedZoneId !== undefined && props.zoneName !== undefined) {
      // TODO: move this back to using constructs
      // prettier-ignore
      // eslint-disable-next-line
      const cfnVpcEndpoint = appNlbComboInterfaceVpcEndpoint.interfaceVpcEndpoint.node.findChild('Resource') as CfnVPCEndpoint;
      // prettier-ignore-end

      const recordSetPrefix = props.recordSetPrefix || 'app-alb-nlb-combo';

      new CfnRecordSet(this, 'EndpointRecordSet', {
        hostedZoneId: props.publicHostedZoneId,
        // FIXME: move to props as string not as import
        name: Fn.join('.', [recordSetPrefix, props.zoneName]),
        type: 'CNAME',
        resourceRecords: [Fn.select(1, Fn.split(':', Fn.select(0, cfnVpcEndpoint.attrDnsEntries)))],
        ttl: '60',
        // Comment, record_set[:comment]
      });

      // FIXTO:
      // new CnameRecord(this, 'cname', {
      //   domainName: endpointdns,
      //   recordName: props.baseHost,
      //   zone: zone
      // }
    }
  }
}
