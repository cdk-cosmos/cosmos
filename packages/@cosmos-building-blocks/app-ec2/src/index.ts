// export {

// } from './app-alb-nlb-combo';

export * from './app-alb-nlb';
