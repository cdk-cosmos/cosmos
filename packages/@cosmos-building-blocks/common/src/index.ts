export { CDKToolkit } from './cdktoolkit';
export { CrossAccountExports, CrossAccountExportsFn, CrossAccountExportsProps } from './cross-account-exports';
