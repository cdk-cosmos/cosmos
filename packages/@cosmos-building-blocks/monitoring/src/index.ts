export { ecsMetric } from './helpers';
export { ClusterDashboard, ClusterDashboardProps } from './dashboard/cluster';
