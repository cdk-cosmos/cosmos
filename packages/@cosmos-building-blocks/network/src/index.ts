export { IResolverRule, ResolverRule, ResolverRuleProps } from './route53/resolver-rule';
export {
  IResolverRuleAssociation,
  ResolverRuleAssociation,
  ResolverRuleAssociationProps,
} from './route53/resolver-rule-association';

export { ITransitGateway, TransitGateway, TransitGatewayProps } from './transit-gateway/transit-gateway';
export {
  ITransitGatewayAttachment,
  TransitGatewayAttachment,
  TransitGatewayAttachmentProps,
  TransitGatewayRoute,
  TransitGatewayRouteProps,
} from './transit-gateway/transit-gateway-attachment';

export { TransparentProxy, TransparentProxyProps } from './internet/transparent-proxy';
