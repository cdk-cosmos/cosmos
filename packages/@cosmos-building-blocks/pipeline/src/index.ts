export { AppNodePipeline, AppNodePipelineProps } from './app-node-pipeline';
