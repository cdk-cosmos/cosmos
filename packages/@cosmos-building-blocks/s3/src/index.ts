// export {

// } from './app-alb-nlb-combo';

export * from './s3';
