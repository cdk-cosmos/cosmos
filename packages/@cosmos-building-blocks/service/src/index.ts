export { EcsService, EcsServiceProps } from './ecs';
