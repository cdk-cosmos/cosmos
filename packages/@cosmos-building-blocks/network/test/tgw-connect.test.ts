import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import '@aws-cdk/core';
import { App, Stack } from '@aws-cdk/core';
import { Vpc, SubnetType, Subnet } from '@aws-cdk/aws-ec2';
import { TransitGateway, TransitGatewayAttachment, ResolverRule, ResolverRuleAssociation } from '../src';

const app = new App();
const testStack = new Stack(app, 'TestStack');
const myvpc = new Vpc(testStack, 'TestVpc', {
  cidr: '10.180.7.0/24',
  maxAzs: 1,
  subnetConfiguration: [
    {
      name: 'Main',
      subnetType: SubnetType.ISOLATED,
      cidrMask: 26,
    },
    {
      name: 'Redis',
      subnetType: SubnetType.ISOLATED,
      cidrMask: 28,
    },
  ],
});

const gateway = TransitGateway.fromGatewayAttributes(testStack, 'TransitGateway', {
  gatewayId: 'tgw-0d4180d3e7c919164',
});
const attachment = gateway.addAttachment('TransitGatewayAttachment', {
  vpc: myvpc,
  subnets: [{ subnetGroupName: 'Main' }],
});
attachment.addRoute('TGWRoute', { destinationCidrBlock: '10.0.0.0/8' });

const resolver = ResolverRule.fromResolverAttributes(testStack, 'ResolverRule', {
  ruleId: 'rslvr-rr-a5c8a7ee5efb41eab',
});
resolver.addAssociation('ResolverRuleAssociation', {
  vpc: myvpc,
});

const testStacksynth = SynthUtils.synthesize(testStack);

describe('Transit Gateway Connection', () => {
  test('should create TGW attachments', () => {
    expect(testStacksynth).toHaveResource('AWS::Route53Resolver::ResolverRuleAssociation');
    expect(testStacksynth).toHaveResource('AWS::EC2::TransitGatewayAttachment');
    expect(testStacksynth).toHaveResource('AWS::EC2::Route');
  });

  test('should match snapshot', () => {
    expect(testStacksynth.template).toMatchSnapshot();
  });
});
