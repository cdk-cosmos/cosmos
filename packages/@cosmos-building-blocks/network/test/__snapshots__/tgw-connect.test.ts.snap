// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transit Gateway Connection should match snapshot 1`] = `
Object {
  "Resources": Object {
    "ResolverRuleResolverRuleAssociation669E300B": Object {
      "Properties": Object {
        "ResolverRuleId": "rslvr-rr-a5c8a7ee5efb41eab",
        "VPCId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::Route53Resolver::ResolverRuleAssociation",
    },
    "TestVpcE77CE678": Object {
      "Properties": Object {
        "CidrBlock": "10.180.7.0/24",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestStack/TestVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVpcMainSubnet1RouteTable1567AA0E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestStack/TestVpc/MainSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcMainSubnet1RouteTableAssociation388FC147": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "TestVpcMainSubnet1RouteTable1567AA0E",
        },
        "SubnetId": Object {
          "Ref": "TestVpcMainSubnet1SubnetC0BB40AC",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcMainSubnet1SubnetC0BB40AC": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.180.7.0/26",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestStack/TestVpc/MainSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Main",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVpcRedisSubnet1RouteTableAssociationAFBE0652": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "TestVpcRedisSubnet1RouteTableBE392024",
        },
        "SubnetId": Object {
          "Ref": "TestVpcRedisSubnet1Subnet1490AFD4",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcRedisSubnet1RouteTableBE392024": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestStack/TestVpc/RedisSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcRedisSubnet1Subnet1490AFD4": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.180.7.64/28",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestStack/TestVpc/RedisSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Redis",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TransitGatewayTransitGatewayAttachment91E0B600": Object {
      "DependsOn": Array [
        "TransitGatewayTransitGatewayAttachmentTGWRouteMainSubnet17BBA6FD6",
      ],
      "Properties": Object {
        "SubnetIds": Array [
          Object {
            "Ref": "TestVpcMainSubnet1SubnetC0BB40AC",
          },
        ],
        "TransitGatewayId": "tgw-0d4180d3e7c919164",
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::TransitGatewayAttachment",
    },
    "TransitGatewayTransitGatewayAttachmentTGWRouteMainSubnet17BBA6FD6": Object {
      "DependsOn": Array [
        "TransitGatewayTransitGatewayAttachment91E0B600",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "10.0.0.0/8",
        "RouteTableId": Object {
          "Ref": "TestVpcMainSubnet1RouteTable1567AA0E",
        },
        "TransitGatewayId": "tgw-0d4180d3e7c919164",
      },
      "Type": "AWS::EC2::Route",
    },
  },
}
`;
